File: sdk/node/test_sdk_checkpoint_export_fixture.mjs
import assert from 'node:assert/strict'
import { mkdtemp, readFile, rm } from 'node:fs/promises'
import { tmpdir } from 'node:os'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import mod from './index.js'

const root = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..')
const fixture = JSON.parse(
  await readFile(path.join(root, 'evaluation', 'sdk-checkpoint-export-v1.json'), 'utf8'),
)

const INLINE_CONFIG = `
default_model = "openai/fixture-model"
providers "openai" {
  apiKey = "fixture-key-never-sent"
  baseUrl = "https://fixture.invalid/v1"
  models "fixture-model" {
    name = "Fixture Model"
  }
}
`.trim()

function assertNoForbiddenFields(value) {
  const forbidden = new Set(fixture.forbidden_fields)
  if (Array.isArray(value)) {
    for (const child of value) {
      assertNoForbiddenFields(child)
    }
    return
  }
  if (value && typeof value === 'object') {
    for (const [key, child] of Object.entries(value)) {
      assert.equal(forbidden.has(key), false, `forbidden diagnostic field ${key}`)
      assertNoForbiddenFields(child)
    }
  }
}

assert.equal(fixture.schema_version, 1)
assert.equal(fixture.fixture_id, 'sdk-checkpoint-export-v1')
for (const field of fixture.required_export_fields) {
  assert.ok(field in fixture.sample_export, `missing export field ${field}`)
}
for (const field of fixture.required_descriptor_fields) {
  assert.ok(field in fixture.sample_export.descriptor, `missing descriptor field ${field}`)
}
assertNoForbiddenFields(fixture.sample_export)

const workspace = await mkdtemp(path.join(tmpdir(), 'a3s-node-sdk-cp-export-'))
try {
  const agent = await mod.Agent.create(INLINE_CONFIG)
  const session = await agent.session(workspace, {
    sessionId: 'node-sdk-checkpoint-export-fixture',
  })
  try {
    assert.equal(typeof session.setSessionCheckpointExportSink, 'function')
    let installed = false
    session.setSessionCheckpointExportSink((exportPayload) => {
      installed = true
      assert.ok(exportPayload.descriptor)
      assert.ok(typeof exportPayload.contentBase64 === 'string')
      return { ok: true }
    }, 5_000)
    assert.equal(installed, false, 'install must not invoke the sink eagerly')
    session.setSessionCheckpointExportSink(null)
  } finally {
    await session.close()
    await agent.close()
  }
} finally {
  await rm(workspace, { recursive: true, force: true })
}

console.log('sdk checkpoint export fixture ok')

File: sdk/python/tests/test_sdk_checkpoint_export_fixture.py
"""Cross-SDK contract checks for live checkpoint export sinks (SDK-CP1)."""

from __future__ import annotations

import asyncio
import json
import tempfile
from pathlib import Path
from typing import Any

from a3s_code import Agent, SessionOptions


FIXTURE = json.loads(
    (
        Path(__file__).resolve().parents[2]
        / "evaluation"
        / "sdk-checkpoint-export-v1.json"
    ).read_text(encoding="utf-8")
)

INLINE_CONFIG = """
default_model = "openai/fixture-model"
providers "openai" {
  apiKey = "fixture-key-never-sent"
  baseUrl = "https://fixture.invalid/v1"
  models "fixture-model" {
    name = "Fixture Model"
  }
}
""".strip()


def _assert_no_forbidden(value: Any) -> None:
    forbidden = set(FIXTURE["forbidden_fields"])
    if isinstance(value, list):
        for child in value:
            _assert_no_forbidden(child)
        return
    if isinstance(value, dict):
        for key, child in value.items():
            assert key not in forbidden, f"forbidden diagnostic field {key}"
            _assert_no_forbidden(child)


def test_sdk_checkpoint_export_fixture() -> None:
    async def scenario() -> None:
        assert FIXTURE["schema_version"] == 1
        assert FIXTURE["fixture_id"] == "sdk-checkpoint-export-v1"
        sample = FIXTURE["sample_export"]
        for field in FIXTURE["required_export_fields"]:
            assert field in sample
        for field in FIXTURE["required_descriptor_fields"]:
            assert field in sample["descriptor"]
        _assert_no_forbidden(sample)

        agent = await Agent.create_async(INLINE_CONFIG)
        with tempfile.TemporaryDirectory(prefix="a3s-python-sdk-cp-export-") as workspace:
            options = SessionOptions()
            options.session_id = "python-sdk-checkpoint-export-fixture"
            session = await agent.session_async(workspace, options)
            try:
                assert hasattr(session, "set_session_checkpoint_export_sink")
                invoked = {"count": 0}

                def handler(export_payload: dict) -> dict:
                    invoked["count"] += 1
                    assert "descriptor" in export_payload
                    assert isinstance(export_payload.get("contentBase64"), str)
                    return {"ok": True}

                session.set_session_checkpoint_export_sink(handler, timeout_ms=5_000)
                assert invoked["count"] == 0
                session.set_session_checkpoint_export_sink(None)
            finally:
                await session.close()
                await agent.close()

    asyncio.run(scenario())
